usrp_drivers/n200/n200_driver.cpp

```diff
@@ -24,271 +24,339 @@
-      stream_args.channels = channels;
-      usrp_d->set_tx_rate(dp.txrate());  // ~450us
-      tx_stream = usrp_d->get_usrp()->get_tx_stream(stream_args);  // ~44ms
-      channels_set = true;
+    DEBUG_MSG("TRANSMIT burst flags: SOB "  << driver_packet.sob() << " EOB "
+      << driver_packet.eob());
+
+    auto stream_setup_begin = std::chrono::steady_clock::now();
+
+    //On start of new sequence, check if there are new USRP channels and if so
+    //set what USRP TX channels and rate(Hz) to use.
+    if (driver_packet.channels_size() > 0 && driver_packet.sob() == true)
+    {
+      DEBUG_MSG("TRANSMIT starting something new");
+      channels = make_tx_channels(driver_packet);
+      stream_args.channels = channels; // REVIEW #15 check that the channels from the driver packet are actually available (a mapping as mentioned in other comment would be handy here)
+                                       // REPLY Need to explain this
+      usrp_d.set_tx_rate(driver_packet.txrate(),channels);  // ~450us REVIEW #15 check that it's not = 0 (default)
+                                                   // REPLY added this check to set_tx_rate
+      tx_stream = usrp_d.get_usrp_tx_stream(stream_args);  // ~44ms REVIEW : idea maybe could we set this up with all channels and then send 0s to any that we don't want to transmit on? then this tx_stream setup could be removed from while loop.
+      usrp_channels_set = true;                                   // REPLY Yes. We will have to see what that looks like on the scope.
     }
-    std::chrono::steady_clock::time_point stream_end = std::chrono::steady_clock::now();
+    auto stream_setup_end = std::chrono::steady_clock::now();
 
-    std::cout << "TRANSMIT stream set up time: "
-      << std::chrono::duration_cast<std::chrono::microseconds>
-                                                  (stream_end - stream_begin).count()
-      << "us" << std::endl;
+    auto time_diff = std::chrono::duration_cast<std::chrono::microseconds>
+                                                  (stream_setup_end - stream_setup_begin).count();
+    DEBUG_MSG("TRANSMIT stream set up time: " << time_diff << "us");
 
 
-    std::chrono::steady_clock::time_point ctr_begin = std::chrono::steady_clock::now();
+    auto ctr_freq_setup_begin = std::chrono::steady_clock::now();
 
-    if (dp.txcenterfreq() > 0) {
-      std::cout << "TRANSMIT center freq " << dp.txcenterfreq() << std::endl;
-      usrp_d->set_tx_center_freq(dp.txcenterfreq(), channels);
+    //If there is new center frequency data, set TX center frequency for each USRP TX channel.
+    if (driver_packet.txcenterfreq() > 0.0)
+    {
+      DEBUG_MSG("TRANSMIT center freq " << driver_packet.txcenterfreq()); // REVIEW #34 should print the actual freq after it's been set (return actual frequency from set_tx_center_freq)
+                                                                                         // REPLY Need to discuss this
```

Comment:
Okay, We think it would be a good check to make sure that the freq was set to the requested freq

Reply:
This is being checked within the usrp.cpp code. The error just doesnt have any handling yet. I feel like this is a good place for a hard failure or some serious consequences since we dont want frequencies changed to something outside our control, and the user probably doesnt want that either. This could either spill into a restricted range, or outside the clrfreq search range for example.

Reply:
I don't think this is a hard failure. This is a matter of tuning the oscillator so it might be fairly common to get an actual frequency that is offset from requested. We could just adjust wave frequency to accommodate this. Requires some testing but I think it might be best to catch this at the earliest point (inside this function)

Reply:
Okay we can expand the functionality here then to do that.